README.md
# (MODX)EvolutionCMS.snippets.ddGetParentId

Gets document parent ID(s) of the required level.


## Requires
* PHP >= 5.6
* [(MODX)EvolutionCMS](https://github.com/evolution-cms/evolution) >= 1.1
* [(MODX)EvolutionCMS.libraries.ddTools](https://code.divandesign.ru/modx/ddtools) >= 0.60


## Installation


### Using [(MODX)EvolutionCMS.libraries.ddInstaller](https://github.com/DivanDesign/EvolutionCMS.libraries.ddInstaller)

Just run the following PHP code in your sources or [Console](https://github.com/vanchelo/MODX-Evolution-Ajax-Console):

```php
//Include (MODX)EvolutionCMS.libraries.ddInstaller
require_once(
	$modx->getConfig('base_path') .
	'assets/libs/ddInstaller/require.php'
);

//Install (MODX)EvolutionCMS.snippets.ddGetParentId
\DDInstaller::install([
	'url' => 'https://github.com/DivanDesign/EvolutionCMS.snippets.ddGetParentId',
	'type' => 'snippet'
]);
```

* If `ddGetParentId` is not exist on your site, `ddInstaller` will just install it.
* If `ddGetParentId` is already exist on your site, `ddInstaller` will check it version and update it if needed.


### Manually


#### 1. Elements → Snippets: Create a new snippet with the following data

1. Snippet name: `ddGetParentId`.
2. Description: `<b>1.4</b> Gets document parent ID(s) of the required level.`.
3. Category: `Core`.
4. Parse DocBlock: `no`.
5. Snippet code (php): Insert content of the `ddGetParentId_snippet.php` file from the archive.


#### 2. Elements → Manage Files

1. Create a new folder `assets/snippets/ddGetParentId/`.
2. Extract the archive to the folder (except `ddGetParentId_snippet.php`).


## Parameters description

* `id`
	* Desctription: Document Id.
	* Valid values: `integer`
	* Default value: `[*id*]` (current document)
	
* `level`
	* Desctription: Parent level.
	* Valid values:
		* `1` — the immediate parent
		* `2` — the parent of the immediate parent
		* `-1` — the last parent
		* `-2` — the parent before the last
		* `integer` — etc
	* Default value: `1`
	
* `result_itemsNumber`
	* Desctription: The number of parents that will be returned.
	* Valid values:
		* `integer`
		* `'all'`
	* Default value: `1`
	
* `result_itemTpl`
	* Desctription: Template for output.  
		Empty items after parsing the template will be ignored (you can calling third-party snippets in the template and set your own display conditions).  
		Available placeholders:
		* `[+id+]` — Parent ID.
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: `'@CODE:[+id+]'`
	
* `result_itemsGlue`
	* Desctription: The string that combines items while rendering.
	* Valid values: `string`
	* Default value: `''`
	
* `result_toPlaceholder`
	* Desctription: Returns value to the placeholder.
	* Valid values:
		* `0`
		* `1`
	* Default value: `0`
	
* `result_toPlaceholder_name`
	* Desctription: Placeholder name.
	* Valid values: `string`
	* Default value: `'ddParent'`


## Examples


### Run the snippet through `\DDTools\Snippet::runSnippet` without DB and eval

```php
//Include (MODX)EvolutionCMS.libraries.ddTools
require_once(
	$modx->getConfig('base_path') .
	'assets/libs/ddTools/modx.ddtools.class.php'
);

//Run (MODX)EvolutionCMS.snippets.ddGetParentId
\DDTools\Snippet::runSnippet([
	'name' => 'ddGetParentId',
	'params' => [
		'level' => -1
	]
]);
```


## Links

* [Home page](https://code.divandesign.ru/modx/ddgetparentid)
* [Telegram chat](https://t.me/dd_code)
* [Packagist](https://packagist.org/packages/dd/evolutioncms-snippets-ddgetparentid)
* [GitHub](https://github.com/DivanDesign/EvolutionCMS.snippets.ddGetParentId)


<link rel="stylesheet" type="text/css" href="https://raw.githack.com/DivanDesign/CSS.ddMarkdown/master/style.min.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.snippets.ddGetParentId changelog


## Version 1.4 (2023-09-04)
* \+ You can just call `\DDTools\Snippet::runSnippet` to run the snippet without DB and eval (see README → Examples).
* \+ README → Installation → Using (MODX)EvolutionCMS.libraries.ddInstaller.
* \+ README_ru, CHANGELOG_ru.
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.60 is required.


## Version 1.3.1 (2020-06-22)
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.40.1 is required (not tested in older versions).
* \* Compatibility with new versions of (MODX)EvolutionCMS.libraries.ddTools.
* \* README, CHANGELOG: Style changes.
* \* README:
	* \- Home.
	* \+ Links.
* \* Composer.json → `require` → `dd/evolutioncms-libraries-ddtools`:
	* \* Renamed from `dd/modxevo-library-ddtools`.
	* \* Version format fixed.


## Version 1.3 (2020-03-02)
* \+ Empty items after parsing `result_itemTpl` will be ignored (you can calling third-party snippets in the template and set your own display conditions).
* \+ Composer.json → Require.


## Version 1.2.2 (2020-02-11)
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.30 is required.
* \* Fixed a bug where `result_itemsNumber` was always equal to `'all'`.
* \* Refactoring and other changes.


## Version 1.2.1 (2018-12-09)
* \* Wrong variable name was fixed.


## Version 1.2 (2017-10-09)
* \+ Added support of the `@CODE:` keyword prefix in the `tpl` parameter.
* \+ Added an ability to return several parents (see the `result_itemsNumber` and `result_itemsGlue` parameters).
* \* The following parameters have been renamed (with backward compatibility):
	* \* `tpl` → `result_itemTpl`.
	* \* `toPlaceholder` → `result_toPlaceholder`.
	* \* `placeholderName` → `result_toPlaceholder_name`.
* \* The snippet result will be returned in anyway (empty string for empty result).
* \* Attention! PHP >= 5.4 is required.
* \* Attention! (MODX)EvolutionCMS >= 1.1 is required.
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.20 is required.


## Version 1.1 (2014-11-05)
* \+ Negative values can now be passed to the `level` parameter for setting parent level from the end (`-1` — the last parent; `-2` — the parent before the last; etc).


## Version 1.0.1 (2013-08-10)
* \* The id of a current document will be returned if it has no parent (root document).


## Version 1.0 (2011-12-18)
* \+ The first release.


<link rel="stylesheet" type="text/css" href="https://raw.githack.com/DivanDesign/CSS.ddMarkdown/master/style.min.css" />
<style>ul{list-style:none;}</style>

src/Snippet.php
<?php
namespace ddGetParentId;

class Snippet extends \DDTools\Snippet {
	protected
		$version = '1.4.0',
		
		$params = [
			//Defaults
			'id' => 0,
			'result_itemsNumber' => 1,
			'result_itemTpl' => '@CODE:[+id+]',
			'result_itemsGlue' => '',
			'result_toPlaceholder' => false,
			'result_toPlaceholder_name' => 'ddParent',
			'level' => 1,
		],
		
		$paramsTypes = [
			'id' => 'integer',
			'result_toPlaceholder' => 'boolean',
		],
		
		$renamedParamsCompliance = [
			'result_itemTpl' => 'tpl',
			'result_toPlaceholder' => 'toPlaceholder',
			'result_toPlaceholder_name' => 'placeholderName',
		]
	;
	
	/**
	 * prepareParams
	 * @version 1.0 (2023-09-04)
	 * 
	 * @param $params {stdClass|arrayAssociative|stringJsonObject|stringHjsonObject|stringQueryFormatted}
	 * 
	 * @return {void}
	 */
	protected function prepareParams($params = []){
		//Call base method
		parent::prepareParams($params);
		
		//ID is not set or invalid
		if ($this->params->id <= 0){
			$this->params->id = \ddTools::$modx->documentIdentifier;
		}
		
		$this->params->result_itemTpl = \ddTools::getTpl($this->params->result_itemTpl);
		
		//For backward compatibility
		if ($this->params->level == 'ultimate'){
			$this->params->level = -1;
		}else{
			$this->params->level = intval($this->params->level);
		}
	}
	
	/**
	 * run
	 * @version 1.0 (2023-09-04)
	 * 
	 * @return {string|object|array}
	 */
	public function run(){
		$result = '';
		
		$parents = \ddTools::getDocumentParentIds([
			'docId' => $this->params->id,
			'level' => $this->params->level,
			'totalResults' => $this->params->result_itemsNumber
		]);
		
		foreach (
			$parents as
			$parentIndex =>
			$parentId
		){
			//Parse item
			$parents[$parentIndex] = \ddTools::parseText([
				'text' => $this->params->result_itemTpl,
				'data' => [
					'id' => $parentId
				]
			]);
			
			//Remove empty items
			if (empty($parents[$parentIndex])){
				unset($parents[$parentIndex]);
			}
		}
		
		$result = implode(
			$this->params->result_itemsGlue,
			$parents
		);
		
		//Если надо, выводим в плэйсхолдер
		if ($this->params->result_toPlaceholder){
			\ddTools::$modx->setPlaceholder(
				$this->params->result_toPlaceholder_name,
				$result
			);
			
			$result = '';
		}
		
		return $result;
	}
}